### PR TITLE
CI: Fix most deprecation warnings from GitHub Actions

### LINKED ISSUE #1074: Migrate CI to use env files for save state
https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

### PR DESCRIPTION
**Pull request by guihkx**
This fixes most deprecation warnings from GitHub Actions.

Changes:

- CI: Replace deprecated `::set-output` command
- CI: Replace deprecated `::save-state` command
- CI: Bump [`dtolnay/rust-toolchain`](https://github.com/dtolnay/rust-toolchain) from `v1` to `master`
- CI: Replace [`Ana06/get-changed-files`](https://github.com/Ana06/get-changed-files) by [`tj-actions/changed-files`](https://github.com/tj-actions/changed-files)

Fixes #1074

### CODE REVIEW

**Comment by guihkx** [comment]
I'll add a changelog in my next rebase.

I'll also try to find a good replacement for the (apparently) unmaintained [Ana06/get-changed-files](https://github.com/Ana06/get-changed-files) action.

**Comment by Alexhuszagh** [comment]
> I'll add a changelog in my next rebase.

The changes are internal so it's not required, since it's just a change to our Github actions.

> For the [dtolnay/rust-toolchain](https://github.com/dtolnay/rust-toolchain) action, I'll also have to use `@master` instead of `@v1`, just to get rid of other `::set-output` warnings. Is that okay with you guys?

I'm personally fine with that until a new release is out, especially since it's dtolnay.

> I'll also try to find a good replacement for the (apparently) unmaintained [Ana06/get-changed-files](https://github.com/Ana06/get-changed-files) action.

Thanks, if you can't find one, any contribution is more than enough. We might have to fork it, change the output variables, pin it, and hope it's minimal work to update again.

**Comment by Alexhuszagh** [comment]
From the [action](https://github.com/cross-rs/cross/actions/runs/3461873354/jobs/5780026657#step:5:29)
> readarray -t added_modified < <(jq -r '.[]' <<<'[\".changes/1135.json\"]')
> parse error: Invalid numeric literal at line 1, column 3

Not sure why this is showing up now, I'm more than happy to add the `no changelog` label once this is merged because that shouldn't be happening.

File: .github/workflows/changelog.yml
Comment (Alexhuszagh) [inline .github/workflows/changelog.yml:17]: Ah this might be the issue. It seems like the quoting in this isn't actually correct.

Comment (guihkx) [reply .github/workflows/changelog.yml:17]: Yeah, `tj-actions/changed-files` seems to escape  `"` by default when we use JSON output. And I didn't find an option to disable it. I could use `sed` to replace `\"` by `"`, but I IMO that's too cumbersome.

Anyway, I've experimented not using JSON output, and instead I'm using the `;` character as the file separator and it worked. I'll push the changes soon.

**Comment by guihkx** [comment]
Okay, I think the changes look good now.

The 'changelog check' job is getting skipped now, but just to make sure it still works, I tested it on on my local repository and [it passed](https://github.com/guihkx/cross/actions/runs/3462754196/jobs/5782060251). ([test commit here](https://github.com/guihkx/cross/commit/0ac3c6f607a656fbbc26e8e47c8aed5e77190e83))

I also renamed the PR because I ended up changing more stuff other than just the `::set-output` warnings.

File: xtask/src/util.rs
Comment (Emilgardis) [inline xtask/src/util.rs:304]: this is not correct, `gha_output` is simply an alias to println/eprintln

Instead we need to write to the file pointed to from `GITHUB_OUTPUT`

**Review by Emilgardis — CHANGES_REQUESTED** [review]

File: .github/workflows/changelog.yml
Comment (Alexhuszagh) [inline .github/workflows/changelog.yml:28]: This change should be fine since we have a very limited subset of characters we accept in file names. Nice.

**Comment by guihkx** [comment]
Okay, I tried to address @Emilgardis's comment to the best of my Rust knowledge.

Hope it isn't that bad... 😅

File: xtask/src/util.rs
Comment (Alexhuszagh) [inline xtask/src/util.rs:303]: If we're returning a result here, this probably can be an `eyre::bail!` case:

```rust
eyre::bail!("output `{tag}` contains newlines, consider serializing with json and deserializing in gha with fromJSON()");
```

**Comment by Alexhuszagh** [comment]
Just that minor nit on the panic, since we return a result now. Otherwise, looks good to me. I'll leave the full review to Emil.

File: xtask/src/util.rs
Comment (Emilgardis) [inline xtask/src/util.rs:421]: this needs to append, not write over.

```suggestion
    let mut file = fs::OpenOptions::new().append(true).open(path)?;
```

**Review by Emilgardis — CHANGES_REQUESTED** [review]
great! Just a minor nit and I think this will be ready.

**Comment by guihkx** [comment]
I addressed the last two comments. Thanks a lot for the reviews, by the way. :)

**Comment by Emilgardis** [comment]
bors try --target *unknown-linux*

**Comment by guihkx** [comment]
Okay, we're down from 1000+ warnings to 478, which is still quite an expressive number.

I think I underestimated a bit the complexity of your CI... 😆 

However, it should be fairly easy to fix the remaining warnings as well.

Let me know if I should do it in this PR, or if I should open a new one.

Thanks!

**Review by Alexhuszagh — APPROVED** [review]
> Let me know if I should do it in this PR, or if I should open a new one.

Either is fine, this looks good to me as-is. I'll let Emil decide and merge, but everything looks sound. If you'd like to fix the rest in a subsequent PR, that would be appreciated as well.

**Comment by Emilgardis** [comment]
I think we can fix the remaining warnings

need to update `docker/setup-buildx-action` to v2 and `docker/login-action` to v2 (unsure if `docker/login-action@v2` would solve it but I think it would)

**Comment by Emilgardis** [comment]
@guihkx I added you as a member in bors so you can run `bors try`

**Comment by guihkx** [comment]
>need to update `docker/setup-buildx-action` to v2 and `docker/login-action` to v2 (unsure if `docker/login-action@v2` would solve it but I think it would)

Yup, updating these two should get rid of most warnings.

There's also `mindsers/changelog-reader-action@v2.0.0`, which is displaying warnings about NodeJS 12, but this issue in specific has been fixed in [v2.2.1](https://github.com/mindsers/changelog-reader-action/releases/tag/v2.2.1). Although a couple of `set-output` warnings [still remain](https://github.com/mindsers/changelog-reader-action/issues/75). 

Nevertheless, I think it should be fairly harmless to just use `@v2` instead of a fixed version.

> @guihkx I added you as a member in bors so you can run `bors try`

Thank you. I'll take this opportunity to check if there are other actions that can be easily updated too.

**Comment by guihkx** [comment]
bors try --target *unknown-linux*

**Comment by guihkx** [comment]
Down to 7 warnings. Nice!

image

The first 2 warnings I'm not sure how to address right now. 🤔 

The next 4 warnings are related to usage of `set-output`, and as I mentioned earlier, come from [here](https://github.com/mindsers/changelog-reader-action/issues/75), and once upstream addresses them, they'll go away naturally since we've switched to `@v2`.

And the very last warning seems harmless to me. It's just GitHub informing that the `macos-latest` runner will soon use macOS 12 instead of 11.

**Comment by Alexhuszagh** [comment]
Only the `set-output` ones are warnings worth dealing with: we use macOS latest intending to use 13 when available, and there's no real reason to check if `rustfmt` is already installed since attempting to install it does so anyway.

**Comment by guihkx** [comment]
Nice! In that case, the PR is ready on my side.

**Review by Emilgardis — APPROVED** [review]
awesome! Thanks a lot!

bors r=Alexhuszagh,emilgardis

**Comment by guihkx** [comment]
Ah, I see...

`readarray` trips when we pass an empty "here string" as argument. For example, this:

```bash
readarray -d ';' -t added_modified <<< ''
echo "array size: ${#added_modified[@]}"
```

Outputs:

```
array size: 1
```

My mistake. In this case, `readarray` isn't the right tool for the job. 

This is easily fixable, though. I'll send a PR soon.

**Pull request merged**
